compiler-rt/lib/profile/InstrProfilingUtil.c

```diff
@@ -258,6 +267,120 @@ COMPILER_RT_VISIBILITY FILE *lprofOpenFileEx(const char *ProfileName) {
   return f;
 }
 
+// Return 1 (true) if the file descriptor Fd represents a file that is on a
+// local filesystem, otherwise return 0.
+COMPILER_RT_UNUSED static int is_local_filesystem(int Fd) {
+#if defined(_AIX)
+  struct statfs Vfs;
+  if (fstatfs(Fd, &Vfs) != 0) {
+    PROF_ERR("%s: fstatfs(%d) failed: %s\n", __func__, Fd, strerror(errno));
+    return 0;
+  }
+
+  int Ret;
+  size_t BufSize = 2048u;
+  char *Buf;
+  int Tries = 3;
+  while (Tries--) {
+    Buf = malloc(BufSize);
+    // mntctl returns -1 if `Buf` is `NULL`.
+    Ret = mntctl(MCTL_QUERY, BufSize, Buf);
+    if (Ret != 0)
+      break;
+    BufSize = *(unsigned int *)Buf;
+    free(Buf);
+  }
+
+  if (Ret != -1) {
+    // Look for the correct vmount entry.
+    char *CurObjPtr = Buf;
+    while (Ret--) {
+      struct vmount *Vp = (struct vmount *)CurObjPtr;
+      _Static_assert(sizeof(Vfs.f_fsid) == sizeof(Vp->vmt_fsid),
+                     "fsid length mismatch");
+      if (memcmp(&Vfs.f_fsid, &Vp->vmt_fsid, sizeof Vfs.f_fsid) == 0) {
+        int Answer = (Vp->vmt_flags & MNT_REMOTE) == 0;
+        free(Buf);
+        return Answer;
+      }
+      CurObjPtr += Vp->vmt_length;
+    }
+  }
+
+  free(Buf);
+  // There was an error in mntctl or vmount entry not found; "remote" is the
+  // conservative answer.
+#endif
+  return 0;
+}
+
+static int isMmapSafe(int Fd) {
+  if (getenv("LLVM_NO_MMAP")) // For testing purposes.
+    return 0;
+#ifdef _AIX
+  return is_local_filesystem(Fd);
+#else
+  return 1;
+#endif
```

Comment:
Wouldn't using `mmap` be an issue even on other platforms when the profile is stored on NFS (e.g. Linux)?

Reply:
I couldn't find any evidence that it's broken, in general, on linux or other platforms. 

Note that in this case, we're using mmap to read a file, while maintaining exclusive access to it (opened with `lprofOpenFileEx`) -- we've seen instances, but on AIX only, where we receive a SIGBUS when accessing mmap'ed memory (in clang, while reading a source file) when the file underneath is changed by another process (happens in a makefile build).

No one reported a problem in the past 5+ years, so I think we're good not doing anything on non-AIX platforms for now.